include/boost/geometry/algorithms/detail/envelope/box.hpp

```diff
@@ -40,36 +42,124 @@ namespace detail { namespace envelope
 {
 
 
-struct envelope_box_on_spheroid
+template
+<
+    std::size_t Index,
+    std::size_t Dimension,
+    std::size_t DimensionCount,
+    typename CS_Tag
+>
+struct envelope_indexed_box
 {
-    template<typename BoxIn, typename BoxOut>
+    template <typename BoxIn, typename BoxOut>
     static inline void apply(BoxIn const& box_in, BoxOut& mbr)
     {
-        BoxIn box_in_normalized = detail::return_normalized<BoxIn>(box_in);
+        detail::indexed_point_view<BoxIn const, Index> box_in_corner(box_in);
+        detail::indexed_point_view<BoxOut, Index> mbr_corner(mbr);
+
+        detail::conversion::point_to_point
+            <
+                detail::indexed_point_view<BoxIn const, Index>,
+                detail::indexed_point_view<BoxOut, Index>,
+                Dimension,
+                DimensionCount
+            >::apply(box_in_corner, mbr_corner);
+    }
+};
 
-        geometry::transform(box_in_normalized, mbr);
+template
+<
+    std::size_t Index,
+    std::size_t DimensionCount
+>
+struct envelope_indexed_box<Index, 0, DimensionCount, spherical_equatorial_tag>
+{
+    template <typename BoxIn, typename BoxOut>
+    static inline void apply(BoxIn const& box_in, BoxOut& mbr)
+    {
+
+       // transform() does not work with boxes of dimension higher
+        // than 2; to account for such boxes we transform the min/max
+        // points of the boxes using the indexed_point_view
+        detail::indexed_point_view<BoxIn const, Index> box_in_corner(box_in);
+        detail::indexed_point_view<BoxOut, Index> mbr_corner(mbr);
+
+        // first transform the units
+        transform_units(box_in_corner, mbr_corner);
```

Comment:
Your arguments are very nice and correct, but I am afraid somehow irrelevant.

Unlike functions, in templates we have partial specialization, which in my mind is the same as being able to define the value of a meta-function for a specific input value.

So what is the general paradigm used here (I have abbreviated some names in order to write less):
```
template <typename Geom, std::size_t Dim, std::size_t DimCount, typename CS_Tag>
struct envelope_algo_implementation
{
    // algo goes here
};

template <typename Geom, typename DimensionCount>
struct envelope_algo_implementation<Geom, 0, DimensionCount, spherical_equatorial_tag>
{
    static inline return_type apply(...)
    {

        // do something special for dimensions 0 and 1 and then call the generic version
        envelope_algo_implementation
            <
                Geom, 2, DimensionCount, spherical_equatorial_tag
            >::apply(...);
    }
};
```
This is the same as having a function that internally does the following check:
```
if (dimension == 0 && tag == spherical_equatorial_tag)
{
    // do something different/special
}
```
According to the design I have followed, I simply indicate how `envelope_algo_implementation` is to be implemented for specific values of template arguments. I do not really see the same dependencies as the ones you propose.

Of course it is indeed possible to move the dependence on `Dimension` one layer down. I personally think it is not necessary, but if this is the preferred choice, I can certainly do it that way.

***Note:*** if we remove the normalization part, then some parts of the envelope code will have to be updated and the design will become even more clear. Dimensions `0` and `1` will be independent from each other and the per-dimension design will make much more sense I think.